### DISCUSSION TITLE
Clean way to shuffle rows/columns of 2D array?

### DISCUSSION THREAD

**Discussion by matzhaugen** [rust-ndarray/ndarray#870] [Q&A]
Is there a clean way to shuffle or sample rows/columns of an Array with/without replacement?

**Comment by bluss** [comment]
This example should really go somewhere better and bigger, but I guess it applies. Sorting along an axis and shuffling is a similar problem - if you use a permutation like in the example https://github.com/rust-ndarray/ndarray/blob/d2375956a96347e5d39aaf0328cbde121170324b/examples/sort-axis.rs

and for sampling, look in ndarray-rand

**Reply by matzhaugen** [reply]
Thanks, this is very helpful. `sample_axis` is what I'm looking for. Would it make sense to implement a version of this method that returns a view instead of copying the data? https://docs.rs/ndarray-rand/0.13.0/ndarray_rand/trait.RandomExt.html#tymethod.sample_axis . This is not essential, but I'm just curious. Often times the resampling is a means to and end which is just to collapse the resampled data onto a single scalar or a lower dimensional space. Something as simple as `let resampled_mean = m.sample_axis(Axis(0), n, SamplingMethod::WithReplacement).iter().axis_mean(Axis(0));` . Thoughts?

This might easily be done by implementing a method called `select_view` on ArrayBase and use it as a replacement [here](https://github.com/rust-ndarray/ndarray/blob/master/ndarray-rand/src/lib.rs#L288) .

A caveat is that this might mess up the stride order and make it slow to iterate over rows/columns.

**Reply by bluss** [reply]
A sampling view is not possible - views can't be random access like that unfortunately. Views can only "select" rectangular grids from the underlying data.

The illustrations [in this article](https://medium.com/better-programming/numpy-illustrated-the-visual-guide-to-numpy-3b1d4976de1d) kind of make the same point, see the text that says *The “view” sign means that no copying is actually done when slicing an array.* and the image above it. The image shows what kinds of rectangular grids there might be.
